From 726d298a538e35151f3c5e66bf4e05e89ec6769a Mon Sep 17 00:00:00 2001
From: Michal Sojka <sojka@os.inf.tu-dresden.de>
Date: Mon, 16 Jan 2012 14:36:54 +0100
Subject: [PATCH] Print panic message also to serial line

---
 src/stdio.cpp |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/stdio.cpp b/src/stdio.cpp
index 8bc3e6c..0f0ea94 100644
--- a/src/stdio.cpp
+++ b/src/stdio.cpp
@@ -38,6 +38,10 @@ void panic (char const *format, ...)
     screen.vprintf (format, args);      // Do not grab lock
     va_end (args);
 
+    va_start (args, format);
+    serial.vprintf (format, args);
+    va_end (args);
+
     Cpu::shutdown();
 }
 
-- 
1.7.7.3

