diff --git a/src/syscall.cpp b/src/syscall.cpp
index b56cecc..7ff66c2 100644
--- a/src/syscall.cpp
+++ b/src/syscall.cpp
@@ -486,7 +486,7 @@ void Ec::sys_assign_gsi()
         sys_finish<Sys_regs::BAD_CAP>();
     }
 
-    r->set_msi (Gsi::set (static_cast<unsigned>(sm->node_base), r->cpu(), r->rid()));
+    r->set_msi (Gsi::set (static_cast<unsigned>(sm->node_base) - NUM_CPU, r->cpu(), r->rid()));
 
     sys_finish<Sys_regs::SUCCESS>();
 }
